packages/gatsby/src/schema/schema.js

```diff
@@ -0,0 +1,362 @@
+  await updateSchemaComposer({
+    schemaComposer,
+    nodeStore,
+    types,
+    thirdPartySchemas,
+    typeMapping,
+    typeConflictReporter,
+    parentSpan,
+  })
+  // const { printSchema } = require(`graphql`)
+  const schema = schemaComposer.buildSchema()
+  // console.log(printSchema(schema))
+  return schema
+}
+
+const rebuildSchemaWithSitePage = ({
+  schemaComposer,
+  nodeStore,
+  typeMapping,
+  typeConflictReporter,
+  parentSpan,
+}) => {
+  const typeComposer = addInferredType({
+    schemaComposer,
+    typeComposer: schemaComposer.getTC(`SitePage`),
+    nodeStore,
+    typeConflictReporter,
+    typeMapping,
+    parentSpan,
+  })
+  processTypeComposer({ schemaComposer, typeComposer, nodeStore, parentSpan })
+  return schemaComposer.buildSchema()
+}
+
+module.exports = {
+  buildSchema,
+  rebuildSchemaWithSitePage,
+}
+
+const updateSchemaComposer = async ({
+  schemaComposer,
+  nodeStore,
+  types,
+  typeMapping,
+  thirdPartySchemas,
+  typeConflictReporter,
+  parentSpan,
+}) => {
+  await addTypes({ schemaComposer, parentSpan, types })
+  await addInferredTypes({
+    schemaComposer,
+    nodeStore,
+    typeConflictReporter,
+    typeMapping,
+    parentSpan,
+  })
+  await addSetFieldsOnGraphQLNodeTypeFields({
+    schemaComposer,
+    nodeStore,
+    parentSpan,
+  })
+  await Promise.all(
+    Array.from(schemaComposer.values()).map(typeComposer =>
+      processTypeComposer({
+        schemaComposer,
+        typeComposer,
+        nodeStore,
+        parentSpan,
+      })
+    )
+  )
+  await addThirdPartySchemas({ schemaComposer, thirdPartySchemas, parentSpan })
+  await addCustomResolveFunctions({ schemaComposer, parentSpan })
+}
+
+const processTypeComposer = async ({
+  schemaComposer,
+  typeComposer,
+  nodeStore,
+  parentSpan,
+}) => {
+  if (
+    typeComposer instanceof schemaComposer.TypeComposer &&
+    typeComposer.hasInterface(`Node`)
+  ) {
+    await addNodeInterfaceFields({ schemaComposer, typeComposer, parentSpan })
+    await addResolvers({ schemaComposer, typeComposer, parentSpan })
+    await addConvenienceChildrenFields({
+      schemaComposer,
+      typeComposer,
+      nodeStore,
+      parentSpan,
+    })
+    await addTypeToRootQuery({ schemaComposer, typeComposer, parentSpan })
+  }
+}
+
+const addTypes = ({ schemaComposer, types, parentSpan }) => {
+  types.forEach(typeOrTypeDef => {
+    if (typeof typeOrTypeDef === `string`) {
+      const addedTypes = schemaComposer.addTypeDefs(typeOrTypeDef)
+      addedTypes.forEach(type => {
+        const { GraphQLInterfaceType, GraphQLUnionType } = require(`graphql`)
+        let typeComposer
+        if (type instanceof GraphQLInterfaceType) {
+          typeComposer = schemaComposer.getOrCreateIFTC(type.name)
+        } else if (type instanceof GraphQLUnionType) {
+          typeComposer = schemaComposer.getOrCreateUTC(type.name)
+        }
+        if (typeComposer) {
+          typeComposer.setResolveType(node => node.internal.type)
```

Comment:
This might not be the best place to do this -- I feel we now have multiple places where we do some kind of initial processing of types? But we can take a second look at this later.
I added `addSchemaMustHaveType` because in the tests the custom abstract types from SDL would be stripped from the built schema otherwise.
And `setResolveType` is similarly needed -- but this means that we only allow custom abstract types that combine types that implement Node. Is that restriction ok? Otherwise we'd have to provide some other way to add a type resolver, but I'm not sureI'd want that.